piv/key.go

```diff
@@ -667,24 +671,91 @@ func (yk *YubiKey) PrivateKey(slot Slot, public crypto.PublicKey, auth KeyAuth)
 	}
 }
 
-type keyECDSA struct {
+// ECDSAPrivateKey is a crypto.PrivateKey implementation for ECDSA
+// keys. It implements crypto.Signer and the method SharedKey performs
+// Diffie-Hellman key agreements.
+//
+// Keys returned by YubiKey.PrivateKey() may be type asserted to
+// *ECDSAPrivateKey, if the slot contains an ECDSA key. See examples
+// on SharedKey.
+type ECDSAPrivateKey struct {
 	yk   *YubiKey
 	slot Slot
 	pub  *ecdsa.PublicKey
 	auth KeyAuth
 	pp   PINPolicy
 }
 
-func (k *keyECDSA) Public() crypto.PublicKey {
+// Public returns the public key associated with this private key.
+func (k *ECDSAPrivateKey) Public() crypto.PublicKey {
 	return k.pub
 }
 
-func (k *keyECDSA) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+var _ crypto.Signer = (*ECDSAPrivateKey)(nil)
+
+// Sign implements crypto.Signer.
+func (k *ECDSAPrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
 	return k.auth.do(k.yk, k.pp, func(tx *scTx) ([]byte, error) {
 		return ykSignECDSA(tx, k.slot, k.pub, digest)
 	})
 }
 
+// SharedKey performs a Diffie-Hellman key agreement with the peer
+// to produce a shared secret key.
+//
+// Peer's public key must use the same algorithm as the key in
+// this slot, or an error will be returned.
+//
+// Length of the result depends on the types and sizes of the keys
+// used for the operation. Callers should use a cryptographic key
+// derivation function to extract the amount of bytes they need.
+func (k *ECDSAPrivateKey) SharedKey(peer crypto.PublicKey) ([]byte, error) {
```

Comment:
Didn't see this addressed in the comments, is there a reason you kept this as crypto.PublicKey instead of *ecdsa.PublicKey? I'd prefer the latter since it enforces type safety.

Reply:
The downside of strong type there is that will never match any future interfaces defined. But that can be sorted if and when such an interface arises.